### ISSUE TITLE
Read a sub-section using get_section and iterate through pages

### ISSUE THREAD

**Issue by NuLL3rr0r** [getzola/zola#250]
From: https://www.getgutenberg.io/documentation/templates/pagination/
> A paginated section gets the same section variable as a normal section page. In addition, a paginated section gets a paginator variable of the Pager type:

I have a few pages inside my content folder and I can iterate through them using paginator.pages. But, if I want to import my old blog from Hexo and keep the current urls which is like this:

```
${my-domain}/blog/${post-tile}
```

Naturally, I'll create a folder named blog inside content and move all the .md files into blog folder, like this:

```
content
    |-- _index.md
    |-- blog
        |-- page1.md
        |-- page2.md
        |-- _index.md
```

Let's say I want content/_index.html as the home page of my blog to read the posts titles from content/blog/ and render them. From inside content/index.md:

```
{% set posts_section = get_section(path="blog/_index.md") %}
```

posts_section.pages will be empty no matter what. And, trying to get the paginator object from that section:

```
{% set post_paginator = post_section.paginator %}
```

Will result in errors since no paginator is available for the section:

```
Reason: Failed to render 'minimal-blog/templates/page.html'
Reason: Variable `post_section.paginator` not found in context while rendering 'minimal-blog/templates/page.html'
```

Is there a way to get a paginator object or pages array from get_section? Or, maybe another way to achieve the same results?

**Comment by Keats** [comment]
The paginator is only in the context when it's rendering the actual section page, not from `get_section`.

> posts_section.pages will be empty no matter what.

Sounds like a bug to me then, do you have an open repository?

For the actual usecase: if you want the site index to be paginated, you will need to put the posts in the root content directory, which means they won't be at `/blog/$slug` but at `/$slug` unless you change the path in each article frontmatter.
Someone asked the same thing a couple of days ago so it might be worth looking into how make that usecase work better. I don't think adding the paginator to `get_section` would work though, the renderer has no way of knowing the page contains pagination and needs to render all pages.

**Comment by NuLL3rr0r** [comment]
Yes, I do have an open repository for that (an unfinished Gutenberg theme). But, I made a few commits afterward and overwrote those changes. I'm not at home at the moment. But, I'll try to find the last commit with the bug mentioned above and post it for your investigation later today.

**Comment by NuLL3rr0r** [comment]
>> posts_section.pages will be empty no matter what.

> Sounds like a bug to me then, do you have an open repository?

I found the commit, but I just figured out it was an issue with my content directory structure which I fixed in the later commits afterward. So, please ignore that one. Sorry for the inconvenience.

Well, I can redirect old /blog/$slug to /$slug as another solution. Although it's not a deal-breaker, it would still be nice to have /blog/$slug since most static generators allow that (at the moment I'm using Hexo and it allows that).

**Comment by Keats** [comment]
>  it would still be nice to have /blog/$slug since most static generators allow that (at the moment I'm using Hexo and it allows that).

How does Hexo handle that?

**Comment by NuLL3rr0r** [comment]
Well, here's how the following static generators handle it:

**Octopress v2**

A global entry in _config.yml:

```
permalink: /blog/:year/:month/:day/:title/
```

**Hexo v3**

A global entry in _config.yml:

```
permalink: blog/:year/:month/:day/:title/
```

**Hugo**

A global entry in config.toml:

```
[permalinks]
  post = "/blog/:year/:month/:day/:title/"
```

It seems pretty standard across various static generators. I'm not sure if this approach works for Gutenberg or not. But, adding /blog to the front matter of every post is counter-productive especially if you need to change that prefix path later on. You have to go through all those files (good old sed/find works in such a case, but still).

**Comment by Keats** [comment]
I think this issue ties in with https://github.com/Keats/gutenberg/issues/84

Being able to say, render the `blog` section at `path = ""` using `template = "index.html"` should solve the issue of using a paginator on another page than its own and the posts urls would stay the same (`blog/${slug}`). This is more flexible than the `permalinks` approach as you can define it on a section basis.
The `:year/:month/:day` pattern wouldn't work but I'm not planning to support that so it's not a big issue.

**Comment by NuLL3rr0r** [comment]
Thanks, I did not know about that issue. I do not care about the :year/:month/:day part. I'll look into your suggestion to see if it works for me. Thanks again.

**Comment by Keats** [comment]
> I'll look into your suggestion to see if it works for me

It's not implemented yet!

**Comment by Keats** [comment]
I was starting to implement it but there's an interesting question: how to deal with conflict? Say I have a section with `path = ""` but also a `content/_index.md`, what should happen? And can someone change the `path` of the index section? 
Gutenberg also automatically creates an index section if there isn't one for simplicity sake so we can assume there always is an index section.

@phil-opp you opened #84, any idea what you would like to see there?

**Comment by NuLL3rr0r** [comment]
Well, good question!

I thought about it a bit. But, a conflict is a conflict; it's an error from the user part. I see two approaches:

1. Silently ignoring the conflict (a warning could be given, too) and writing the pages at that location (which obviously overwrites the content at that path by the latest page that gets generated)
2. Giving an error about the conflict and stop

Regarding the ability to the path of the index section, I don't see a point in changing that. I'm not sure if I have a use case for that myself.

**Comment by Keats** [comment]
1 is not desirable: sections are rendered in parallel so ordering is not guaranteed.

I'm wondering if @phil-opp usecase is the same as this issue and whether a `render_on_index` attribute on a section makes more sense than allowing to render paths at random locations.

**Comment by maghoff** [comment]
I'm not sure whether to comment on this issue or create a new one. Let's try commenting on this one.

I have a similar use case, where I want the root index page to act as the index page for _multiple_ sections, so I'd like to get the paginator for a given section.

I have `{% set blog = get_section(path="blog/_index.md") -%}`, which lets me iterate over `blog.pages`, which is neat, but has two problems:

 1. it does not act on the pagination configuration (but I can manually take a slice)
 2. it gives me all the pages, even the ones that have `draft = true` (but I can filter with `if`)

Note that the workarounds I presented do not compose well; if I take a slice of _n_ elements and then filter with an `if`-condition inside the loop, I might end up with fewer than _n_ elements rendered.

A couple of solutions I imagine could be better:

 a. A way to get the paginator object for the first page of the given section. I imagine either `get_section(…).paginator` or `get_paginator(…)`. Something along these lines.
 b. A new filter to filter out the drafts: `blog.pages | discard_drafts | slice(end=5)`

**Comment by Keats** [comment]
A bit late but I think we could generalise solution b in tera by adding a `filter` filter

**Comment by yaymukund** [comment]
Is there a better way to do this today? Would you accept a PR that implements `get_paginator()`, [as was suggested](https://github.com/Keats/gutenberg/issues/250#issuecomment-379393102)?

I don't want my root directory (`public/`) to be littered with dozens of directories like `public/my-post-1-slug/`, `public/my-post-2-slug/`, etc. Ideally, my posts wouldn't live alongside (e.g.) `robots.txt`, `.well-known/`, and other stuff that *needs* to be at the root by design.

**Comment by Keats** [comment]
There is now a `filter` filter in Tera so you can do `{{ posts | filter(attribute="draft", value=true)  | slice(end=5) }}`

**Comment by yaymukund** [comment]
@Keats thanks, I was able to fudge it to work: https://github.com/yaymukund/yaymukund-blog/blob/97f98e9e34b7fe4d3b5a356c524ef6ab9d3207d3/templates/index.html#L24 :+1: 

Thank you for maintaining this library. I'm a fan :)

**Comment by Keats** [comment]
Thanks for the kind words!

@NuLL3rr0r I believe this could be accomplished by #408 ? Same usecase after all.

**Comment by Keats** [comment]
`transparent` sections are now in the `next` branch if you want to play with them
